internal/extender/failover.go

```diff
@@ -0,0 +1,296 @@
+			pod.DeletionTimestamp != nil ||
+			pod.Spec.NodeName == "" ||
+			podsWithRRs[pod.Name] {
+			// count only spark scheduler managed pods without resource reservations
+			continue
+		}
+		appID := pod.Labels[SparkAppIDLabel]
+		sp, ok := appIDToPods[appID]
+		if !ok {
+			sp = &sparkPods{}
+		}
+		switch pod.Labels[SparkRoleLabel] {
+		case Driver:
+			sp.driver = pod
+		case Executor:
+			sp.executors = append(sp.executors, pod)
+		default:
+			svc1log.FromContext(ctx).Error("received non spark pod, ignoring", svc1log.SafeParams(internal.PodSafeParams(*pod)))
+		}
+		appIDToPods[appID] = sp
+	}
+	return appIDToPods
+}
+
+func availableResourcesPerInstanceGroup(
+	ctx context.Context,
+	rrs []*v1beta1.ResourceReservation,
+	nodes []*v1.Node,
+	overhead resources.NodeGroupResources) (map[string]resources.NodeGroupResources, map[string][]*v1.Node) {
+	sort.Slice(nodes, func(i, j int) bool {
+		return nodes[j].CreationTimestamp.Before(&nodes[i].CreationTimestamp)
+	})
+
+	schedulableNodes := make(map[string][]*v1.Node)
+	for _, n := range nodes {
+		if n.Spec.Unschedulable {
+			continue
+		}
+		instanceGroup := n.Labels[instanceGroupNodeSelector]
+		schedulableNodes[instanceGroup] = append(schedulableNodes[instanceGroup], n)
+	}
+	usages := resources.UsageForNodes(rrs)
+	usages.Add(overhead)
+	availableResources := make(map[string]resources.NodeGroupResources)
+	for instanceGroup, ns := range schedulableNodes {
+		availableResources[instanceGroup] = resources.AvailableForNodes(ns, usages)
+	}
+	return availableResources, schedulableNodes
+}
+
+// syncResourceReservation gets a stale resource reservation and updates its status to reflect all given executors
+func (r *reconciliator) patchResourceReservation(execs []*v1.Pod, rr *v1beta1.ResourceReservation) error {
+	for _, e := range execs {
+		for name, reservation := range rr.Spec.Reservations {
+			if reservation.Node != e.Spec.NodeName {
+				continue
+			}
+			currentPodName, ok := rr.Status.Pods[name]
+			if !ok {
+				rr.Status.Pods[name] = e.Name
+				break
+			}
+			pod, err := r.podLister.Pods(e.Namespace).Get(currentPodName)
+			if errors.IsNotFound(err) || (err == nil && isPodTerminated(pod)) {
+				rr.Status.Pods[name] = e.Name
+				break
+			}
+		}
+	}
+	return r.resourceReservations.Update(rr)
+}
+
+func (r *reconciliator) constructResourceReservation(
+	ctx context.Context,
+	driver *v1.Pod,
+	executors []*v1.Pod,
+	instanceGroup string) (*v1beta1.ResourceReservation, resources.NodeGroupResources, error) {
+	applicationResources, err := sparkResources(ctx, driver)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	nodes, nodesFound := r.orderedNodes[instanceGroup]
+	availableResources, resourcesFound := r.availableResources[instanceGroup]
+	if !nodesFound || !resourcesFound {
+		return nil, nil, werror.Error("instance group not found", werror.SafeParam("instanceGroup", instanceGroup))
+	}
+
+	executorCountToReserve := applicationResources.executorCount - len(executors)
```

Comment:
Don't you want to create reservations for stale executors instead?

Reply:
I think the wording is off here, these executors do not exist yet so we need to reserve spaces for them, as opposed to `executors` here that exist but do not have the corresponding resource reservation object